### ISSUE TITLE
Suggestion: public `sf::Image` constructors should be converted to factory functions

### ISSUE THREAD

**Issue by vittorioromeo** [SFML/SFML#3158]
### Prerequisite Checklist

- [X] I searched for [existing issues](https://github.com/search?q=repo%3ASFML%2FSFML&type=issues) to prevent duplicates
- [X] I searched for [existing discussions on the forum](https://www.google.com/search?q=site%3Ahttps%3A%2F%2Fen.sfml-dev.org) to prevent duplicates
- [X] I understand [SFML's scope](https://www.sfml-dev.org/contribute.php#general-considerations) and believe the feature fits it

### Describe your feature request here

Currently we have:

```cpp
class Image 
{
    // ...
    explicit Image(const Vector2u& size, const Color& color = Color::Black);
    Image(const Vector2u& size, const std::uint8_t* pixels);
    // ...
};
```

These two constructors can "fail" as they can construct an image in an invalid state, if *any* of these conditions apply:
- `size.x == 0`
- `size.y == 0`
- `pixels == nullptr`

Throughout `sf::Image`'s implementation, these conditions are checked repeatedly. E.g.:

```cpp
// ...
[[nodiscard]] bool Image::copy(const Image& source, const Vector2u& dest, const IntRect& sourceRect, bool applyAlpha)
{
    // Make sure that both images are valid
    if (source.m_size.x == 0 || source.m_size.y == 0 || m_size.x == 0 || m_size.y == 0)
        return false;
// ...
```

I suggest that `sf::Image` object should be "always valid" thanks to `std::optional`-based construction:

```cpp
class Image 
{
    // ...
    [[nodiscard]] static std::optional<Image> create(const Vector2u& size, const Color& color = Color::Black);
    [[nodiscard]] static std::optional<Image> create(const Vector2u& size, const std::uint8_t* pixels);
    // ...
};
```

If any of the three "failure" conditions mentioned above would arise here, `std::nullopt` would be returned.

This change would:
- Ensure that every `sf::Image` object is always in a valid state (non-zero size, with available pixels)
- Simplify/optimize the implementation of `sf::Image`, as there would be no need to actively check for invalid states
- Synergize well with #3157 
- Simplify `sf::Image` tests, e.g. https://github.com/SFML/SFML/blob/master/test/Graphics/Image.test.cpp#L189-L195

I can provide a PR if there is agreement on this suggestion.

---

You can see an implementation of this change in my personal fork of SFML:

- [`Image.hpp` L69-L92](https://github.com/vittorioromeo/SFML/blob/wip_nonglobal_graphicscontext/include/SFML/Graphics/Image.hpp#L69-L92)

- [`Image.cpp` L104-L148](https://github.com/vittorioromeo/SFML/blob/wip_nonglobal_graphicscontext/src/SFML/Graphics/Image.cpp#L104-L148)

**Comment by ChrisThrasher** [comment]
https://github.com/SFML/SFML/pull/3152 handles this by adding `sf::Image::Image()` and explicitly defining the empty state of an image. This solution is consistent with many other resource management types in SFML which are inherently `optional`-y. 

If #3152 is not merged then I do support converting these constructors to factory functions and removing this internal empty state from `sf::Image`.

**Comment by ChrisThrasher** [comment]
Closed by #3152 because are no longer using factory functions.
